### ISSUE TITLE
Allow Checkout Error Placement to Be Easily Moved

### ISSUE THREAD

**Issue by spencerfinnell** [easydigitaldownloads/easy-digital-downloads#5517]
Semi related to #2222. I have a PR ready that allows the anchor point of where errors are inserted to be more easily changed.

https://github.com/easydigitaldownloads/easy-digital-downloads/blob/master/assets/js/edd-ajax.js#L392

This can be filtered easily and an added event on a checkout error allows themes/plugins to extend what happens next even more. 

Is this something you all would be open to in the meantime before #2222?

**Comment by pippinsplugins** [comment]
@spencerfinnell Yes! I would love to have this.

**Comment by spencerfinnell** [comment]
@pippinsplugins Brought my PR up to date. Any chance this could get milestoned?

**Comment by pippinsplugins** [comment]
@spencerfinnell milestoned for 2.8!

**Comment by SDavisMedia** [comment]
@spencerfinnell I tested the filter and got mixed results. It definitely works:

```php
function whatever( $content ) {
	$content['checkout_error_anchor'] = '#edd_cc_address';
	return $content;
}
add_filter( 'edd_global_checkout_script_vars', 'whatever' );
```
https://cl.ly/3E0o250P3D1D

Doesn't appear to work with all gateways, though. Stripe, for example, continues to use the old default error anchor.

Words from Pippin:
<img width="431" alt="screenshot 2017-06-07 11 02 31" src="https://user-images.githubusercontent.com/2359131/26893674-daa197ec-4b70-11e7-9c1d-2504213d525c.png">

@pippinsplugins does something have to happen in all gateways to take advantage of this enhancement?

**Comment by pippinsplugins** [comment]
@SDavisMedia Stripe and 2Checkout should be the only gateways that need updated. They both add errors to the checkout via their own JS. We just need to update those to target this new anchor.

**Comment by mintplugins** [comment]
Will this only work for default cart errors or for any error set using `edd_set_error`?

Also, should there be a fallback in case the ID entered in the function is not present on the page? For example, in @SDavisMedia example, he hooked it to `#edd_cc_address`, but no such field exists if you are purchasing in test mode.

**Comment by mintplugins** [comment]
Confirmed that it only works for default cart errors. I suppose that is fine because that will be addressed in https://github.com/easydigitaldownloads/easy-digital-downloads/issues/2222 correct?

**Comment by cklosowski** [comment]
So since gateways are going to have to be updated...is our status that this is fully backwards compatible when not updating the gateways immediately?

**Comment by cklosowski** [comment]
@spencerfinnell @mintplugins

**Comment by mintplugins** [comment]
@cklosowski It is fully backwards compatible as-is. Errors will still show in the same place by default. 

The only thing I'd still like to see would be a fallback so that if the anchor element doesn't exist or is hidden, the errors still show. However, I suppose it could be argued that making sure this works in those scenarios would be up to the person filtering the anchor placement.

**Comment by spencerfinnell** [comment]
Yeah the JS could check for the existence of https://github.com/easydigitaldownloads/easy-digital-downloads/pull/5518/files#diff-28089abe109462083feccaa41af37d26R392 first and fallback to a default. However like @mintplugins said if you're already making the change via the filter it's kind of up to you to ensure you make it an element that exists.

**Comment by cklosowski** [comment]
As long as at release no gateways break and errors are still output I'm fine. We can go and update each gateway to use the new "feature" later.

**Comment by SDavisMedia** [comment]
If we're not concerned about the fallback for when the target element doesn't exist, all is well on my end. The filter works as it should and gateways that can't use it will still output errors in the default spot. :+1:

**Comment by cklosowski** [comment]
After seeing some odd behavior on the EDD site, @pippinsplugins and I are reopening this. It appeared to have caused some issues with the Stripe Gateway and error handling which, when used in combination with auto register was causing a validation issue.

**Comment by cklosowski** [comment]
Ok I got this back into `release/2.8` could use some testing. the issues were caused by another change in the JS files. I was just reverting all of them to start and working my way back through them to find the culprit.

**Comment by davidsherlock** [comment]
I feel this ticket needs to be re-opened, as issues still exist with Stripe and other payment gateways in relation to the placement of the error messages.

Here is the related customer ticket: https://secure.helpscout.net/conversation/611576065/85129?folderId=1847245

**Comment by SDavisMedia** [comment]
@davidsherlock I don't think so. This issue addresses the ability to move the errors in EDD core, and is backwards compatible. If there are gateway-specific concerns, as mentioned through the issue, they should be addressed in the gateways.

**Comment by nosilver4u** [comment]
Just for clarity, what I discussed with David in that ticket was the following scenario:
When using Stripe on the checkout page, Stripe puts it's error messages in an appropriate place, but does not use the filter mentioned. EDD core errors follow the filter as described above in this issue, but NOT when using Stripe. When Stripe is in use, the checkout data is submitted via POST, instead of JS/AJAX, and the EDD core errors end up in an odd place, which requires a customer to scroll all the way to the bottom of the checkout page to even find out what the error was (if they ever DO scroll, and don't just give up on the checkout). It might be appropriate to open two separate issues, one for the Stripe errors being movable, and another for refining the EDD core errors when Stripe is active.
